aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,661 @@
+    /// signer.
+
+    inline fun map_or<PermKey: copy + drop + store, T>(
+        permissioned: &signer,
+        perm: PermKey,
+        mutate: |&mut StoredPermission| T,
+        default: T,
+    ): T {
+        let permission_signer_addr = permission_address(permissioned);
+        assert!(
+            exists<PermissionStorage>(permission_signer_addr),
+            error::permission_denied(E_NOT_ACTIVE)
+        );
+        let perms =
+            &mut borrow_global_mut<PermissionStorage>(permission_signer_addr).perms;
+        let key = copyable_any::pack(perm);
+        if (simple_map::contains_key(perms, &key)) {
+            mutate(simple_map::borrow_mut(perms, &key))
+        } else {
+            default
+        }
+    }
+
+    inline fun insert_or<PermKey: copy + drop + store>(
+        permissioned: &signer,
+        perm: PermKey,
+        mutate: |&mut StoredPermission|,
+        default: StoredPermission,
+    ) {
+        let permission_signer_addr = permission_address(permissioned);
+        assert!(
+            exists<PermissionStorage>(permission_signer_addr),
+            error::permission_denied(E_NOT_ACTIVE)
+        );
+        let perms =
+            &mut borrow_global_mut<PermissionStorage>(permission_signer_addr).perms;
+        let key = copyable_any::pack(perm);
+        if (simple_map::contains_key(perms, &key)) {
+            mutate(simple_map::borrow_mut(perms, &key));
+        } else {
+            simple_map::add(perms, key, default);
+        }
+    }
+
+    /// Authorizes `permissioned` with a given capacity.
+    ///
+    /// Consumption using `check_permission_consume` will deduct the capacity.
+    public fun authorize<PermKey: copy + drop + store>(
+        master: &signer,
+        permissioned: &signer,
+        capacity: u256,
+        perm: PermKey
+    ) acquires PermissionStorage {
+        assert!(
+            is_permissioned_signer(permissioned)
+                && !is_permissioned_signer(master)
+                && signer::address_of(master) == signer::address_of(permissioned),
+            error::permission_denied(ECANNOT_AUTHORIZE)
+        );
+        insert_or(
+            permissioned,
+            perm,
+            |stored_permission| {
+                increase_capacity(stored_permission, capacity);
+            },
+            StoredPermission::Capacity(capacity),
+        )
+    }
+
+    /// Authorizes `permissioned` with the given unlimited permission.
+    /// Unlimited permission can be consumed however many times.
+    public fun authorize_unlimited<PermKey: copy + drop + store>(
+        master: &signer,
+        permissioned: &signer,
+        perm: PermKey
+    ) acquires PermissionStorage {
+        assert!(
+            is_permissioned_signer(permissioned)
+                && !is_permissioned_signer(master)
+                && signer::address_of(master) == signer::address_of(permissioned),
```

Comment:
will `signer::address_of(permissioned)` return the address of the master?

Reply:
That's correct. That's how we make this feature entirely backwards compatible with our eixsitng modules on chain.